### PR TITLE
9: Create Elevator Subsystem

### LINKED ISSUE #9: Create Elevator Subsystem
The elevator subsystem will contain two motors driven in sync. The code will need to have several base functions available such as:
- Automatic Zeroing based upon limit switch input
- Software hard-stop to prevent overextension.
- Move to position. (Needs to be fast.)
- Variable percent input control (Basic control for stuff like climbing or operator control)

Once these base functions are properly implemented we will have to do an analysis of the mechanism and its possible interactions with other subsystems to come up with a good state-machine.

### PR DESCRIPTION
**Pull request by coolinm**
I promise i made the base develop this time

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -29,6 +29,7 @@ public void robotInit()
   public void robotPeriodic() 
   {
     CommandScheduler.getInstance().run();
+
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/Robot.java]: whitespace?

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -82,7 +83,7 @@ public void teleopInit()
   @Override
   public void teleopPeriodic() 
   {
-    // Intentionally Empty
+
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/Robot.java]: // Intentionally Empty

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -56,7 +69,9 @@ private void configureBindings()
                 () -> -m_primaryController.getRightX()
       ));
 
-      m_primaryController.a().onTrue(new SeedFieldRelative(m_Drivetrain));
+      m_primaryController.a().whileTrue(new SetElevatorPosition(m_Elevator, .1));
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/RobotContainer.java]: We don't need your testing stuff

File: src/main/java/frc/robot/commands/elevator/SetElevatorPosition.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  public void execute()
+  {
+    // Intentionally Empty
+  }
+
+  // Called once the command ends or is interrupted.
+  @Override
+  public void end(boolean interrupted)
+  {
+    m_Elevator.setElevatorPercent(0);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/commands/elevator/SetElevatorPosition.java]: Don't set to zero

File: src/main/java/frc/robot/constants/Hardware.java [hunk 5]
```diff
@@ -29,4 +30,9 @@ public class Hardware
     public static final int SHOOTER_MOTOR_1_ID = 9;
     public static final int SHOOTER_MOTOR_2_ID = 10;
     public static final int WRIST_MOTOR_ID = 11;
+
+    // Elevator
+    public static final int ELEVATOR_CANCODER_ID = 5;
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/constants/Hardware.java]: move this down

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  private final TalonFX m_ElevatorMotor1 = new TalonFX(Hardware.ELEVATOR_MOTOR_1_ID);
+  private final TalonFX m_ElevatorMotor2 = new TalonFX(Hardware.ELEVATOR_MOTOR_2_ID);
+  private final CANcoder m_CANcoder = new CANcoder(Hardware.ELEVATOR_CANCODER_ID);
+
+  /** Creates a new Elevator. */
+  public Elevator()
+  {
+    // Apply gains to both motors
+    m_ElevatorMotor1.getConfigurator().apply(ElevatorConstants.ELEVATOR_GAINS);
+    m_ElevatorMotor2.getConfigurator().apply(ElevatorConstants.ELEVATOR_GAINS);
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Elevator.java]: only need one gains >:(

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  public StatusSignal<Double> getElevatorPosition()
+  {
+    return m_ElevatorMotor1.getPosition();
+  }
+
+  /**
+   * @brief Gets the CANcoder's current absolute position.
+   * @return A Status Signal of CANcoder's current absolute position.
+   */
+  public StatusSignal<Double> getCANcoderPosition()
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Elevator.java]: CANCoder

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  {
+    return m_CANcoder.getAbsolutePosition();
+  }
+
+  @Override
+  public void periodic()
+  {
+    SmartDashboard.putNumber("Elevator pos", this.getElevatorPosition().getValue());
```
Comment (jasonsoohoo) [inline src/main/java/frc/robot/subsystems/Elevator.java]: put this on a timer

**Review by jasonsoohoo — CHANGES_REQUESTED** [review]
I require changes

**Comment by coolinm** [comment]
Changes made @jasonsoohoo

**Review by jasonsoohoo — APPROVED** [review]

**Review by cjahner1 — APPROVED** [review]
Good job colin

**Pull request merged**